File: 2015/day05/day05.exs
defmodule Day05Part1 do
  def run(path) do
    IO.puts "Part 1: " <> to_string(nice_count(lines(path)))
  end

  def nice_count(lines), do: Enum.count(lines, &nice?/1)
  def nice?(line), do: three_vowels?(line) && !contains_sequence?(line) && repeating_letter?(line)
  def three_vowels?(line), do: Regex.scan(~r/[aeiou]/i, line) |> length >= 3
  def contains_sequence?(line), do: String.contains?(line, ["ab", "cd", "pq", "xy"])
  def repeating_letter?(line), do: Regex.match? ~r/(.)\1/, line
  defp lines(path), do: File.read!(path) |> String.split("\n", trim: true)
end

# defmodule Day05Part2 do
#   def run(path) do
#     IO.inspect nice_count(lines(path))
#
#     IO.inspect lines(path)
#   end
#
#   def nice_count(lines), do: Enum.count(lines, &nice?/1)
#   def nice?(line), do: three_vowels?(line) && !contains_sequence?(line) && repeating_letter?(line)
#   def three_vowels?(line), do: Regex.scan(~r/[aeiou]/i, line) |> length >= 3
#   def contains_sequence?(line), do: String.contains?(line, ["ab", "cd", "pq", "xy"])
#   def repeating_letter?(line), do: Regex.match? ~r/(.)\1/, line
#   defp lines(path), do: File.read!(path) |> String.split("\n", trim: true)
# end

Day05.run("day05.input")

File: 2015-ruby/lib/day_05.rb
require 'digest'

class Day05
  def part1(input)
    prepare_words(input).count{ |word| nice?(word) }
  end

  def nice?(word) 
    three_vowels?(word) && double_letter?(word) && !bad_sequence?(word)
  end

  def three_vowels?(word)
    word.scan(/[aeiou]/).length >= 3
  end

  def double_letter?(word)
    word.match?(/(.)\1/)
  end

  def bad_sequence?(word)
    ['ab', 'cd', 'pq', 'xy'].any?{ |str| word.include?(str) }
  end

  def prepare_words(input) input.split end

  # def part2(input)
  # end
end
